com.unity.ml-agents/Runtime/Inference/TensorGenerator.cs

```diff
@@ -95,29 +95,36 @@ public void InitializeObservations(List<ISensor> sensors, ITensorAllocator alloc
         {
             // Loop through the sensors on a representative agent.
             // For vector observations, add the index to the (single) VectorObservationGenerator
-            // For visual observations, make a VisualObservationInputGenerator
+            // For visual observations, make a NonVectorObservationInputGenerator
             var visIndex = 0;
             VectorObservationGenerator vecObsGen = null;
             for (var sensorIndex = 0; sensorIndex < sensors.Count; sensorIndex++)
             {
                 var sensor = sensors[sensorIndex];
                 var shape = sensor.GetObservationShape();
                 // TODO generalize - we currently only have vector or visual, but can't handle "2D" observations
-                var isVectorSensor = (shape.Length == 1);
-                if (isVectorSensor)
+                var rank = shape.Length;
+                switch (rank)
                 {
-                    if (vecObsGen == null)
-                    {
-                        vecObsGen = new VectorObservationGenerator(allocator);
-                    }
+                    case 1:
+                        if (vecObsGen == null)
+                        {
+                            vecObsGen = new VectorObservationGenerator(allocator);
+                        }
+                        vecObsGen.AddSensorIndex(sensorIndex);
+                        break;
+                    case 2:
+                        m_Dict[TensorNames.ObservationPlaceholderPrefix + sensorIndex] =
+                        new NonVectorObservationInputGenerator(sensorIndex, allocator);
+                        break;
+                    case 3:
```

Comment:
Is there a reason why we need to keep using `TensorNames.VisualObservationPlaceholderPrefix + visIndex` while `TensorNames.ObservationPlaceholderPrefix + sensorIndex` also gives a unique name? Is it to preserve compatibilities for old models?

Reply:
Spot on! Yes, it is for backwards compatibility, we can change all obs to be TensorNames.ObservationPlaceholderPrefix + sensorIndex at the next major release